service/qos/service_level_controller.cc

```diff
@@ -86,15 +86,26 @@ void service_level_controller::set_distributed_data_accessor(service_level_distr
     }
 }
 
-future<> service_level_controller::stop() {
-    // unregister from the service level distributed data accessor.
-    _sl_data_accessor = nullptr;
-    if (this_shard_id() == global_controller) {
-        // abort the loop of the distributed data checking if it is running
-        _global_controller_db->dist_data_update_aborter.request_abort();
-        _global_controller_db->notifications_serializer.broken();
+future<> service_level_controller::drain() {
+    if (this_shard_id() != global_controller) {
+        return make_ready_future();
     }
-    return std::exchange(_distributed_data_updater, make_ready_future<>());
+    // abort the loop of the distributed data checking if it is running
+    _global_controller_db->dist_data_update_aborter.request_abort();
+    _global_controller_db->notifications_serializer.broken();
+    return _global_controller_db->distributed_data_update.then_wrapped([] (future<> f) {
+        try {
+            f.get();
+        } catch (const broken_semaphore& ignored) {
+        } catch (const sleep_aborted& ignored) {
+        } catch (const exceptions::unavailable_exception& ignored) {
+        } catch (const exceptions::read_timeout_exception& ignored) {
+        }
+    });
+}
+
+future<> service_level_controller::stop() {
+    return make_ready_future<>();
```

Comment:
If start aborts due to an exception before `service_level_controller::drain` is deferred, then the whole stop becomes a no-op.

Reply:
Right, that's no good

Reply:
I guess I should either make draining idempotent and call it in stop() or move the updater initialization right before deferring the drain. I'll go with making it idempotent.